RFC: drivers: biometrics: Add driver subsystem, GT5x, and ZFM-X0 fingerprint sensor drivers

heronet opened the pull request:
This PR proposes a new `biometrics` driver subsystem to standardise interactions with biometric sensors (Fingerprint, Iris, Face, etc.).

Currently, Zephyr lacks a dedicated API for these devices. Users are forced to use raw UART/I2C, which is not portable. This RFC introduces a generic API and a reference implementation.

This subsystem provides a unified API for:
- Enrollment (Start, Capture, Finalize)
- Matching (Verify 1:1, Identify 1:N)
- Template Management (Store, Delete, List)
- Feedback (LED control)


On the naming of the (Biometrics) API:

> A biometric sensor is a device that measures and analyzes biological characteristics or behavioral patterns of individuals for the purpose of identifying or verifying their identity. These sensors capture unique physical or behavioral traits that are distinct to each person, allowing for accurate authentication.

(source: https://ieeexplore.ieee.org/document/10808205) (thanks to @rodrigopex)

More references:
1. [ISO/IEC 2382-37](https://www.iso.org/standard/73514.html): International standard for "Biometric recognition"
2. Android SDK uses [android.hardware.biometrics](https://developer.android.com/reference/android/hardware/biometrics/package-summary) for its authentication system

pdgendt commented on drivers/biometrics/biometrics_as608.c:
Use the `DEVICE_API` macro to place it in a dedicated linker region (generated automatically). This is useful for filtering devices, for example in shell commands.

```suggestion
static DEVICE_API(biometric, as608_api) = {
```

pdgendt reviewed (commented):
Cool stuff! This will need to go through the architecture working group as it introduces a new driver API class.

Some nice additions would be:
- Documentation!
- Maintainer area
- A simulated/emulated test
- Shell commands for testing

kartben commented on drivers/biometrics/Kconfig.as608:
```diff
@@ -0,0 +1,29 @@
+
+if BIOMETRICS_AS608
+
+config AS608_PASSWORD
```
This should probably be in Devicetree since the password is unique to each sensor?

kartben commented on samples/drivers/biometrics/fingerprint/README.rst:
```diff
@@ -0,0 +1,90 @@
+##########################
+
+Overview
+********
```
this needs to use `.. zephyr:code-sample::` (template here https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/templates/sample.tmpl or see other code samples for reference)

heronet commented on include/zephyr/drivers/biometrics.h:
```diff
@@ -0,0 +1,726 @@
+
+	if (api->get_capabilities == NULL) {
+		return -ENOSYS;
+	}
```
Agreed. I've categorized the APIs and removed the NULL checks for mandatory operations

rodrigopex commented:
@heronet 

Could you try adding another sensor driver implementation to this API to ensure the APIs are sufficiently generic? It would be ideal to have at least one of each device available in the enum. 

The name seems to be okay: 

> A biometric sensor is a device that measures and analyzes biological characteristics or behavioral patterns of individuals for the purpose of identifying or verifying their identity. These sensors capture unique physical or behavioral traits that are distinct to each person, allowing for accurate authentication.

(source: https://ieeexplore.ieee.org/document/10808205)

heronet committed:
drivers: biometrics: introduce biometrics subsystem

Introduces a new drivers subsystem for biometric sensors.
The API supports enrollment, matching, and template storage.

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
doc: biometrics: Add API documentation

Add overview and API reference documentation for the biometrics
driver subsystem.

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
dts: bindings: add biometrics binding

Adds device tree binding for Zhiantec ZFM-X0 fingerprint
sensor and biometrics-emul

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
drivers: biometrics: add ZFM-x0 fingerprint sensor driver

Add driver for ZFM-x0 series optical fingerprint sensors supporting
ZFM-X0, R30x, JM-101, DY-50, FPM10A, and compatible modules.

Features:
- Multi-sample enrollment flow (2 samples)
- 1:1 verification and 1:N identification matching
- LED control
- Configurable security level and timeout

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
samples: drivers: biometrics: add fingerprint sample

Adds a sample application demonstrating the Biometrics
subsystem with a fingerprint sensor.

The sample demonstrates:
- Getting sensor capabilities
- Two-stage enrollment process
- Verification (1:1 matching)
- Identification (1:N matching)
- Template management

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
drivers: biometrics: add emulated driver for testing

Add emulated biometrics driver for testing shell commands
and application logic. Implements full API:
- Configurable attributes
- Template storage with simulated enrollment workflow
- Match operations with configurable success simulation
- LED state tracking

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
drivers: biometrics: add shell commands for testing

Add shell interface for biometrics driver subsystem providing
commands for:
- info: display device capabilities
- attr_get/attr_set: read/write device attributes
- enroll: full enrollment flow and individual steps
- template_list/delete/delete_all: template management
- match: verify or identify operations
- led: LED control

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
tests: drivers: biometrics: Add emulated driver test

- Add test suite for biometrics API using emulated driver
- Test enrollment flow: start, capture, finalize, abort
- Test template operations: store, read, delete, list
- Test matching: verify and identify modes
- Test attribute get/set and error handling
- Runs on native_sim and real hardware

Signed-off-by: Siratul Islam <email@sirat.me>

heronet commented:
> > @nordicjm After doing some deep diving I found out that the (almost) exact same fingerpring firmware and AS608 chip is used in JM101, Grow R307, Adafruit Optical Fingerprint(DY-50, FPM10A), ZhianTec ZFM 20/60/70, DFrobot Optical Fingerprint. But the original seems to be ZFM as both the Adafruit and DFRobot datasheets links to this one.
> > I think `synochip,as608` is not the most accurate compatible afterall. `zhiantec,zfm-x0` sounds the most accurate to me. So I changed it to this one.
> > Adafruit: https://cdn-shop.adafruit.com/datasheets/ZFM%20user%20manualV15.pdf DFRobot: https://dfimg.dfrobot.com/nobody/wiki/f4fd9de3830b858d4c661a6f12acf305.pdf ZFM-70: https://www.electrokit.com/upload/quick/01/6c/77db_41022692_TDS.pdf Grow R307: https://www.openhacks.com/uploadsproductos/r307_fingerprint_module_user_manual.pdf
> 
> `zhiantec,zfm-70` would be fine for dts compatible (assuming it is the one you in your device), as that string should refer to one specific device, e.g. lis2dh does the same then checks if the device has other compatibles to see if it should enabled quriks so https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/seeed/wio_wm1110_dev_kit/wio_wm1110_dev_kit.dts#L97 has both compatibles - the lis2dh is required for the driver, the lis3dh enables various extra parts in the driver https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/st/lis2dh/lis2dh.c#L591

Well I have DY-50/FPM10A, JM-101, and Grow R307 and all of these are compatible with this driver. The ZFMs are the same according to the datasheets. I beleive the relevant features are all the same so using seperate compatible wouldn't be necessary.

Should I go with `grow,r307` or `zhiantec,zfm-70`?

heronet committed:
dts: bindings: biometrics: add ADH GT5X fingerprint sensor

Add device tree binding for ADH Technology GT5X series optical
fingerprint sensors (GT-511C1R, GT-511C3, GT-521F32, GT-521F52).

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
drivers: biometrics: add GT5X fingerprint sensor driver

Add driver for ADH Technology GT5X series optical fingerprint
sensors. Supports GT-511C1R, GT-511C3, GT-521F32, and GT-521F52
variants.

Features:
- 3-stage enrollment with retry support
- 1:1 verification and 1:N identification
- LED control
- Model-agnostic design via device tree configuration

Signed-off-by: Siratul Islam <email@sirat.me>

heronet committed:
MAINTAINERS: Add biometrics drivers subsystem entry

Add maintainer entry for the new biometrics drivers subsystem,
covering drivers, bindings, headers, samples, and tests

Signed-off-by: Siratul Islam <email@sirat.me>

heronet commented:
Hi @nashif @henrikbrixandersen @carlescufi @pdgendt @rodrigopex @fabiobaltieri @bjarki-andreasen @JarmouniA,

I've addressed all the review feedback. It's been a couple weeks without movement. 

Is there anything else I need to do to move this forward? Happy to make any changes or additions needed.

Thanks!

pdgendt commented:
> I've addressed all the review feedback. It's been a couple weeks without movement.

I think its due to being so close to the holidays, it will probably get more reviews from next week onwards.

heronet commented:
Hi @pdgendt @rodrigopex @fabiobaltieri @nashif @henrikbrixandersen,

Hope the new year is going well! It's been a while since the holidays..
Would appreciate any further feedback when you have the bandwidth.

Thanks!

bjarki-andreasen reviewed (dismissed):
Still think the LED pattern part should not be part of this API, but overall it looks good

heronet commented on tests/drivers/biometrics/biometrics_emul/src/main.c:
```diff
@@ -0,0 +1,340 @@
+void biometrics_emul_set_match_fail(const struct device *dev, bool fail);
+void biometrics_emul_set_capture_timeout(const struct device *dev, bool timeout);
+void biometrics_emul_set_image_quality(const struct device *dev, int quality);
+enum biometric_led_state biometrics_emul_get_led_state(const struct device *dev);
```
@bjarki-andreasen I moved the headers to `include/zephyr/drivers/biometrics/emul.h`.

mbolivar reviewed (changes requested):
Nice work! Please add a rationale for your default values or mark them as `required`, then I'm happy to approve this PR.
